engine/servermaster/server.go

```diff
@@ -525,32 +453,94 @@ func (s *Server) initElector() error {
 }
 
 func (s *Server) serve(ctx context.Context) error {
+	var (
+		cleaupFuncs []func()
+		cleanupOnce sync.Once
+	)
+	cleanup := func() {
+		cleanupOnce.Do(func() {
+			for _, f := range cleaupFuncs {
+				f()
+			}
+		})
+	}
+	defer cleanup()
+
 	errGroup, ctx := errgroup.WithContext(ctx)
 
 	// TODO: Support TLS.
 	tcpServer, err := tcpserver.NewTCPServer(s.cfg.Addr, &security.Credential{})
 	if err != nil {
 		return err
 	}
-	defer tcpServer.Close()
+	cleaupFuncs = append(cleaupFuncs, func() {
+		if err := tcpServer.Close(); err != nil {
+			log.Warn("failed to close tcp server", zap.Error(err))
+		}
+	})
 	errGroup.Go(func() error {
 		return tcpServer.Run(ctx)
 	})
 
-	httpServer, err := s.createHTTPServer()
+	grpcServer := s.createGRPCServer()
+	cleaupFuncs = append(cleaupFuncs, grpcServer.Stop)
+	errGroup.Go(func() error {
+		return grpcServer.Serve(tcpServer.GrpcListener())
+	})
+
+	// gRPC-Gateway doesn't call gRPC interceptors when it directly forwards requests to the service handler.
+	// See https://github.com/grpc-ecosystem/grpc-gateway/issues/1043.
+	//
+	// To make the gRPC interceptors work, we have to create a client connection to the gRPC server and register it to
+	// the gRPC-Gateway ServerMux. bufconn is used to create an in-process gRPC server, so the client connection can
+	// bypass the network stack. See https://github.com/grpc/grpc-go/issues/906 for more details.
+	bufConn := bufconn.Listen(grpcConnBufSize)
+	cleaupFuncs = append(cleaupFuncs, func() {
+		if err := bufConn.Close(); err != nil {
+			log.Warn("failed to close bufconn", zap.Error(err))
+		}
+	})
+	inProcessGRPCServer := s.createGRPCServer()
+	cleaupFuncs = append(cleaupFuncs, inProcessGRPCServer.Stop)
+	errGroup.Go(func() error {
+		return inProcessGRPCServer.Serve(bufConn)
+	})
+
+	dial := func(context.Context, string) (net.Conn, error) { return bufConn.Dial() }
+	conn, err := grpc.DialContext(
+		ctx,
+		"bufnet",
+		grpc.WithContextDialer(dial),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cleaupFuncs = append(cleaupFuncs, func() {
+		if err := conn.Close(); err != nil {
+			log.Warn("failed to close grpc connection", zap.Error(err))
+		}
+	})
+
+	httpServer, err := s.createHTTPServer(conn)
 	if err != nil {
 		return err
 	}
-	defer httpServer.Close()
+	cleaupFuncs = append(cleaupFuncs, func() {
+		if err := httpServer.Close(); err != nil {
+			log.Warn("failed to close http server", zap.Error(err))
+		}
+	})
 	errGroup.Go(func() error {
 		return httpServer.Serve(tcpServer.HTTP1Listener())
 	})
 
-	grpcServer := s.createGRPCServer()
-	defer grpcServer.Stop()
 	errGroup.Go(func() error {
-		return grpcServer.Serve(tcpServer.GrpcListener())
+		<-ctx.Done()
+		cleanup()
+		return errors.Trace(ctx.Err())
 	})
```

Comment:
Is it necessary to check `ctx.Done()` here, other error group must have checked it. If this block can be removed, `cleanup` doesn't need to be protected with `cleanupOnce`

Reply:
Some background goroutines may still be running after context is canceled. We need to explicitly stop or close them and wait for them to exit.